apps/confidential/confidential.go

```diff
@@ -0,0 +1,364 @@
+	key, err := x509.ParsePKCS8PrivateKey(der)
+	if err != nil {
+		return nil, fmt.Errorf("problems decoding private key using PKCS8: %w", err)
+	}
+	return key, nil
+}
+
+// Credential represents the credential used in confidential client flows.
+type Credential struct {
+	secret string
+
+	cert *x509.Certificate
+	key  crypto.PrivateKey
+}
+
+// toMSALBASE returns the accesstokens.Credential that is used internally. The current structure of the
+// code requires that client.go, requests.go and confidential.go share a credential type without
+// having import recursion. That requires the type used between is in a shared package. Therefore
+// we have this.
+// TODO(jdoak): change method name.
+func (c Credential) toMSALBASE() *accesstokens.Credential {
+	return &accesstokens.Credential{Secret: c.secret, Cert: c.cert, Key: c.key}
+}
+
+// NewCredFromSecret creates a Credential from a secret.
+func NewCredFromSecret(secret string) (Credential, error) {
+	if secret == "" {
+		return Credential{}, errors.New("secret can't be empty string")
+	}
+	return Credential{secret: secret}, nil
+}
+
+// NewCredFromCert creates a Credential from an x509.Certificate and a PKCS8 DER encoded private key.
+// CertFromPEM() can be used to get these values from a PEM file storing a PKCS8 private key.
+func NewCredFromCert(cert *x509.Certificate, key crypto.PrivateKey) Credential {
+	return Credential{cert: cert, key: key}
+}
+
+// Client is a representation of authentication client for confidential applications as defined in the
+// package doc. A new Client should be created PER SERVICE USER.
+// For more information, visit https://docs.microsoft.com/azure/active-directory/develop/msal-client-applications
+type Client struct {
+	client.Base
+
+	cred *accesstokens.Credential
+
+	// userID is some unique identifier for a user. It actually isn't used by us at all, it
+	// simply acts as another hint that a confidential.Client is for a single user.
+	userID string
+}
+
+// Options are optional settings for New(). These options are set using various functions
+// returning Option calls.
+type Options struct {
+	// Accessor controls cache persistence.
+	// By default there is no cache persistence. This can be set using the Accessor() option.
+	Accessor cache.ExportReplace
+
+	// The host of the Azure Active Directory authority.
+	// The default is https://login.microsoftonline.com/common. This can be changed using the
+	// Authority() option.
+	Authority string
+}
+
+func (o Options) validate() error {
+	u, err := url.Parse(o.Authority)
+	if err != nil {
+		return fmt.Errorf("the Authority(%s) does not parse as a valid URL", o.Authority)
+	}
+	if u.Scheme != "https" {
+		return fmt.Errorf("the Authority(%s) does not appear to use https", o.Authority)
+	}
+	return nil
+}
+
+// Option is an optional argument to New().
+type Option func(o *Options)
+
+// Authority allows you to provide a custom authority for use in the client.
+func Authority(authority string) Option {
```

Comment:
A common pattern I've seen is to prefix the variadic config funcs with a "With" prefix.  It's a nice visual cue that this is for configuration, plus all the options are sorted together in Intellisense.

Reply:
I'm not sure it is common enough pattern to add clarity, but it also isn't obtrusive.  Maybe it will become common enough.

I've added With to all options in confidential and public.